keps/0021-kudo-upgrade.md

```diff
@@ -0,0 +1,271 @@
+title: Upgrades of KUDO installation
+authors:
+  - "@aneumann82"
+owners:
+  - "@aneumann82"
+editor: TBD
+creation-date: 2019-11-28
+last-updated: 2019-12-02
+status: provisional
+---
+
+# Upgrading KUDO
+
+## Table of Contents
+<!--ts-->
+   * [Upgrading KUDO](#upgrading-kudo)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+         * [Current State](#current-state)
+      * [Open Questions](#open-questions)
+      * [Proposal](#proposal)
+         * [KUDO Prerequisites](#kudo-prerequisites)
+            * [Proposal for update process](#proposal-for-update-process)
+         * [KUDO Manager](#kudo-manager)
+            * [Proposal for update process](#proposal-for-update-process-1)
+         * [CRDs](#crds)
+            * [Proposal for update process](#proposal-for-update-process-2)
+         * [KUDO CLI](#kudo-cli)
+            * [Proposal for update process](#proposal-for-update-process-3)
+      * [Updating KUDO installation](#updating-kudo-installation)
+         * [Upgrade Steps](#upgrade-steps)
+      * [User Stories](#user-stories)
+            * [Story 1](#story-1)
+            * [Story 2](#story-2)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations TODO](#risks-and-mitigations-todo)
+            * [Failure cases](#failure-cases)
+      * [Graduation Criteria TODO](#graduation-criteria-todo)
+      * [Drawbacks TODO [optional]](#drawbacks-todo-optional)
+      * [Alternatives](#alternatives)
+      * [Infrastructure Needed](#infrastructure-needed)
+      * [Resources](#resources)
+      * [Implementation History](#implementation-history)
+
+<!-- Added by: aneumann, at: Mon Dec  2 11:34:32 CET 2019 -->
+
+<!--te-->
+
+## Summary
+
+TBD
+
+## Motivation
+
+We strive for quick and regular updates of KUDO. We need a process for upgrading all the moving parts of KUDO, and how 
+the different parts interact, and what kind of compatibility we want to provide.
+
+### Goals
+
+- How updates of KUDO are executed
+  - Updates of CRDs
+  - Updates of the KUDO manager
+  - Updates of prerequisites (Namespaces, RoleBindings, ServiceAccounts, etc.)
+- Interoperability
+  - How and if multiple versions of CRDs are maintained
+  - How a version of KUDO CLI work with older and newer CRD versions
+- How to handle operators that are not supported by a new KUDO version
+
+### Non-Goals
+
+- Updates/Upgrades of operators and operator versions itself
+- Versioning of KUDO itself
+- Multi-Tenancy, i.e. running multiple KUDO installations on the same cluster
+
+
+### Current State
+
+At the moment, KUDO does not provide any migration capabilities and needs a clean installation to use a new version.
+
+## Open Questions
+- Lowest supported K8s version
+- Do we want to support downgrades? 
+- Split versions between KUDO manager and KUDO CLI?
```

Comment:
They're basically two different programs that only communicate via CRDs. It might make something easier having split versions, for example bug fixes in manager only without having a new KUDO CLI version that does not change anything in the CLI.

But it would introduce a new compatibility matrix, and having "empty" version changes on manager or CLI isn't that big of an issue. I lean towards having a single version for both parts of KUDO.

Reply:
yeah, me too I think until there's a compelling reason to split them, having them as one is a big win, I would almost go as far as resolve this question with no :) especially since there's no real as for this and the scope of this KEP is already pretty big